### DISCUSSION TITLE
Setting global request timeout

### DISCUSSION THREAD

**Discussion by PakToPack** [roadrunner-server/roadrunner#1607] [Q&A] [answered]
Hello! I configured all my scripts etc. But now I'm facing a problem:
Sometimes requests take longer than 300ms and, ofcourse, that is loading my workers heavily with 1000 rps.

Can I somehow break the connection after 300 ms and return 204 error? Really need that

**Comment by rustatian** [comment]
Hey @PakToPack 👋🏻 

Break connection? Do you mean some DB connection or ?

**Reply by PakToPack** [reply]
No, I mean this kind of behaviour:
Client sent request to our server
This request incoming to roadrunner server.
If the request inside the worker is in-time of 300 ms - respond 200 and response
If worker took longer than 300ms - finish the job with 204 error.

So I need some kind of hard-limit of 300ms for overall request processing.

If it is less - it's okay.

Cuz I kinda finished the project and it happens, that CURL (Guzzle HTTP in this case is a wrapper) can slow down it's response when I hit some level of requests. So I need server-side limit to close the connection after 300ms.

**Reply by rustatian** [reply]
What is the problem of sending the `204` response code and setting timeout from the PHP?

From the RR side you may set this code to 204: https://github.com/roadrunner-server/roadrunner/blob/master/.rr.yaml#L597 (but this response code would be sent on any error)
And set `exec_ttl` timeout to `300ms`: https://github.com/roadrunner-server/roadrunner/blob/master/.rr.yaml#L882 (should be like `400ms`, because all system calls to the PHP process are counted. FYI, if you exceed this timeout, RR will kill the worker.

**Reply by rustatian** [reply]
Generally this is not a good practice to use like `400ms` timeout and kill workers after that. Because most likely, since you have >1000RPS, you'll get even more timeouts, because RR would have to reallocate killed workers.

**Reply by rustatian** [reply]
You can also write a simple middleware in Go, connect it to the RR, catch the response, perform the needed HTTP call and populate the response structure. You can also check timings, set the time on request, check it back on response `after handler.Serve(r,w)` Go function. In this case you would have full control over the HTTP request and can also set any response code you want.
Tutorial: https://roadrunner.dev/docs/customization-middleware/current/en

**Reply by PakToPack** [reply]
> Generally this is not a good practice to use like `400ms` timeout and kill workers after that. Because most likely, since you have >1000RPS, you'll get even more timeouts, because RR would have to reallocate killed workers.

Yep, but idk what else to do.
With >1000 rps you get a chance that half of connections are using workers and cant let them happen. Idk if that a curl issue or a server issue or php issue.
The thing is: I have 256 workers now, that ARE doin' fine.

Also I have a test script that utilizes GuzzleHTTP (for cURL concurrent requests) that is sending 4 requests to different servers each every connection.

When this cURL solution is not in production, but just in testing script, where I personally do like 2-3 requests per second - all good. When all 1000 rps is coming to that same root, it reacts unexpected:
First requests are loading at a speed of 300-500 ms. Then 2000ms, then 5000 ms and after like 2-3 minutes server is crashing to error code 1/failed.

Idk what to do, cuz my app is based on contacting other servers by HTTP protocol. Overall all these tests standalone are great, in production usage they ain't worker.

So IDK if it is a cURL or php curl problem, but our whole production stopped because we are trin' to solve the problem and what can be wrong.

If I'd count, it feels like 4000 (if we get 1000 rps) outgoing cURL requests per second is crazy. But that's what we need. And we can't find a solution, except of killing script.

**Reply by rustatian** [reply]
If you have a lot of blocking HTTP calls, plus they are independent and could take advantage of parallelization, use the middleware written in GO (link above). When you make a request in the middleware, the worker is free at that moment and can process the next request while you're doing the jobs in the middleware. You can send the necessary values for the outgoing HTTP requests in the response headers and delete them afterwards (to avoid exposing them to the client).

**Reply by rustatian** [reply]
If you have the requirements (logic, etc) for this middleware, I may be able to help you with it as it looks quite small.

**Reply by PakToPack** [reply]
> If you have the requirements (logic, etc) for this middleware, I may be able to help you with it as it looks quite small.

The thing is: I need to pickup data from middleware. I need the data to process request. Not just use an api

**Reply by rustatian** [reply]
Middleware works in both ways. Before and After request.

**Reply by PakToPack** [reply]
> Middleware works in both ways. Before and After request. 

Can you provide a link to that middleware? Iunno what is that and how to use it. I'm a PHP dev, not GoLang 😅

**Reply by rustatian** [reply]
--> https://github.com/orgs/roadrunner-server/discussions/1607#discussioncomment-6187122

**Reply by PakToPack** [reply]
> --> https://github.com/orgs/roadrunner-server/discussions/1607#discussioncomment-6187122

Hello. So I got a question:
Can I interact with this middleware in the middle of php worker execution? Just because before I'm parsing results from http requests, I need to get into the database

**Reply by rustatian** [reply]
You can create a regular plugin with the RPC interface and offload any tasks to it at any point in your PHP code.

**Reply by PakToPack** [reply]
> You can create a regular plugin with the RPC interface and offload any tasks to it at any point in your PHP code.

Oh, ok, I'll dive into that. Thx 🙏

**Reply by PakToPack** [reply]
Hi! I've got a problem with my local WSL server.
I can't built roadrunner with Velox, I get this:
```bash
2023-06-20T12:31:01.005+0300    DEBUG   [STDERR OUTPUT] {"log": "go: found github.com/roadrunner-server/informer/v4 in github.com/roadrunner-server/informer/v4 v4.1.9\ngo: found github.com/roadrunner-server/resetter/v4 in github.com/roadrunner-server/resetter/v4 v4.0.6\ngo: found github.com/roadrunner-server/config/v4 in github.com/roadrunner-server/config/v4 v4.3.4\ngo: found github.com/roadrunner-server/rpc/v4 in github.com/roadrunner-server/rpc/v4 v4.1.11\ngo: found github.com/minio/highwayhash in github.com/minio/highwayhash v1.0.2\ngo: found github.com/nats-io/jwt/v2 in github.com/nats-io/jwt/v2 v2.4.1\n"}
2023-06-20T12:31:01.043+0300    INFO    [CHECKING OUTPUT DIR]   {"dir": "/home/eugene/rr_built"}
2023-06-20T12:31:01.061+0300    INFO    [EXECUTING CMD] {"cmd": "/usr/local/go/bin/go build -v -trimpath -ldflags -s -X github.com/roadrunner-server/roadrunner/v2/internal/meta.version= -X github.com/roadrunner-server/roadrunner/v2/internal/meta.buildTime= -o /tmp/roadrunner-server-v2023.1.0-rc.1/roadrunner-server-roadrunner-1c718eb/rr cmd/rr/main.go"}
2023-06-20T12:31:01.291+0300    DEBUG   [STDERR OUTPUT] {"log": "github.com/roadrunner-server/roadrunner/v2023/container\n"}
2023-06-20T12:31:01.348+0300    DEBUG   [STDERR OUTPUT] {"log": "github.com/roadrunner-server/roadrunner/v2023/internal/cli/serve\n"}
2023-06-20T12:31:01.390+0300    DEBUG   [STDERR OUTPUT] {"log": "github.com/roadrunner-server/roadrunner/v2023/internal/cli\n"}
2023-06-20T12:31:01.418+0300    DEBUG   [STDERR OUTPUT] {"log": "command-line-arguments\n"}
2023-06-20T12:31:02.097+0300    INFO    [MOVING EXECUTABLE]     {"file": "/tmp/roadrunner-server-v2023.1.0-rc.1/roadrunner-server-roadrunner-1c718eb/rr", "to": "/home/eugene/rr_built/rr"}
2023-06-20T12:31:02.103+0300    INFO    [CLEANING UP]   {"file/folder": "/tmp/roadrunner-server-v2023.1.0-rc.1"}
2023-06-20T12:31:02.107+0300    INFO    [CLEANING UP]   {"file/folder": "/tmp/roadrunner-server-v2023.1.0-rc.1.zip"}
2023-06-20T12:31:02.107+0300    FATAL   [BUILD FAILED]  {"error": "open /home/eugene/rr_built/rr: permission denied"}
```
Execution command:
```bash
./vx build -c velox.toml -o /home/eugene/rr_built
```
Go and System type:
![image](https://github.com/roadrunner-server/roadrunner/assets/37382014/91a9d5b6-27de-46cc-9b33-18e9e0e53fb7)

**Reply by rustatian** [reply]
What is the reason to use `2023.1.0-rc.1` ?

**Reply by PakToPack** [reply]
> What is the reason to use `2023.1.0-rc.1` ?

Test built purposes. I just got a default config

**Reply by rustatian** [reply]
Don't use `RC` releases for tests. As for the error, it clearly states that velox `./vx` don't have permissions to write to that folder: `open /home/eugene/rr_built/rr: permission denied`

**Reply by PakToPack** [reply]
> Don't use `RC` releases for tests. As for the error, it clearly states that velox `./vx` don't have permissions to write to that folder: `open /home/eugene/rr_built/rr: permission denied`

Is there some kind of instruction on building rr with my custom Plugin and calling it with PHP code through RPC? Cuz I think, that I'm doing smthn wrong. In video tutorial also said "We will discuss more complicated functionality of plugins later" and didn't do so. I have a clean GoLang script for now, all I need - pass json array of links to it through RPC and return back the JSON array of results I got with those links.
But now I'm frustrated, cuz I don't see further details on how to do that.

**Reply by rustatian** [reply]
Here you are: https://github.com/roadrunner-php/goridge#example

**Reply by rustatian** [reply]
`App.Hi` mentioned in the example is here: https://github.com/roadrunner-server/goridge#sample-of-usage

**Reply by PakToPack** [reply]
> `App.Hi` mentioned in the example is here: https://github.com/roadrunner-server/goridge#sample-of-usage

Ok, so I shouldn't implement this in velox? Just use like in Example. Server.go, registering App there and method, like makeRequest, pass my JSON in that method through RPC call in client.php and return a new JSON with results from Server.go? Without implementing it to roadrunner through velox, right?

**Reply by rustatian** [reply, marked as answer]
No, you've seen the video-tutorial. You need to create an RPC interface on your `plugin.go` with methods you need.
`Velox` is used to build RR with your custom plugin. If you wanted to build it by yourself you may add your plugin to the list of the [RR plugins](https://github.com/roadrunner-server/roadrunner/blob/master/container/plugins.go) and build the server like a regular Go app.
Then, you may use this:
```php
$rpc->call("App.Hi", "Antony");
```
To call the methods from your plugin, where, `App` is the name of your plugin, `Hi` is the method you're calling and `Antony` is the argument you passed to your plugin. It's better to use [`protobuf`](https://protobuf.dev/reference/php/php-generated/) as the argument you passing to the plugin. You may have a look at the `roadrunner-php/jobs` PHP part and `roadrunner-server/jobs` Go part.

**Reply by PakToPack** [reply]
Ok, so the last question for now is:
I didn't ever work with GoLang. So I see in examples things called "go.mod, go.sum". How should I generate them?

**Reply by rustatian** [reply]
1. `go mod init <your module name>`, then:
2. [Tutorial 1](https://go.dev/doc/tutorial/getting-started)
3. [Tutorial 2](https://gobyexample.com/)

**Reply by PakToPack** [reply]
> `Velox` is used to build RR with your custom plugin. If you wanted to build it by yourself you may add your plugin to the list of the [RR plugins](https://github.com/roadrunner-server/roadrunner/blob/master/container/plugins.go) and build the server like a regular Go app.
> Then, you may use this:

So should I add something like this in my velox.toml?
```toml
make_request = { ref = "master", owner = "roadrunner-server", repository = "plugin_sample", replace = "/var/www/roadrunner/plugins"}
```

**Reply by rustatian** [reply]
There is no `plugin_sample` repository. Create repository you own, then replace with your local copy. Use you/you_org as the `owner`.
